Enabling of parallelization of `analysis.hydrogenbonds.WaterBridgeAnalysis`

talagayev opened the pull request:
<!-- 
Insert issue number that this PR fixes (if any) just after 'Fixes #'.
If this PR does not fix an existing issue, consider opening one or remove 'Fixes #' from the PR description.
-->
Fixes #4666.

Changes made in this Pull Request:
<!-- Describe the changes that this PR makes. If applicable, use the following bullet list. --> 
- Addition of parallelization to `WaterBridgeAnalysis ` class in `analysis.hydrogenbonds.wbridge_analysis`.
- Addition of `client_WaterBridgeAnalysis` to `conftest.py`.
- Adjustment of tests in `test_wbridge.py` to work with parallelization.

Here currently I adjusted `analysis.hydrogenbonds.wbridge_analysis` to work with parallelization and added an example of pytests that will show that it works.

The problem that I have encountered is that I had to modify `universe_DA_PBC`, since if I directly add the `client_WaterBridgeAnalysis` it will led to an infinite pytest on my local PC, mainly connected with us using here `StringIO` that appearantly has difficulties with parallelization.

The second thing is that currently the tests in `test_wbridge.py` are mainly with 1 frame in mind, so to showcase that parallelization works I could add additional tests similar like `universe_DA_PBC_10frames()`.

## PR Checklist
<!-- Please use the following checklist to ensure the PR is ready to be reviewed/merged. -->
 - [x] Issue raised/referenced?
 - [ ] Tests updated/added?
 - [ ] Documentation updated/added?
 - [ ] `package/CHANGELOG` file updated?
 - [x] Is your name in `package/AUTHORS`? (If it is not, add it!)

## Developers Certificate of Origin
<!-- 
In order for this PR to be merged you **must certify that you are able to submit your code to be included in MDAnalysis**.
-->

I certify that I can submit this code contribution as described in the [**Developer Certificate of Origin**](https://developercertificate.org/), under the MDAnalysis [LICENSE](https://github.com/MDAnalysis/mdanalysis/blob/develop/LICENSE).


<!-- readthedocs-preview mdanalysis start -->
----
📚 Documentation preview 📚: https://mdanalysis--5151.org.readthedocs.build/en/5151/

<!-- readthedocs-preview mdanalysis end -->

talagayev committed:
added parallelization to WaterBridgeAnalysis

talagayev commented:
@marinegor I would ping you in this PR, since it is parallelization :)

Here I would need your opinion on how to approach the pytest cases. I didn't add `client_WaterBridgeAnalysis` to the tests with `StringIO` since these would lead to the tests being infinite. The question is, is there any goaround for parallelization with `StringIO` or do we just modify the pytests to work?

Also as mentioned with the pytests being only for 1 frame, should I add for each test cases with multiple frames and test them as in the example that I have now or what would be the prefereable approach?

marinegor commented:
@talagayev I don't see a reason why these files take `StringIO` instead of just a filename, hence I'd suggest re-writing these tests to accept filenames, and put respective filenames (`ADA.gro`, `DWA.gro`, ...) in e.g. `testsuite/MDAnalysisTests/data` or in a new folder `testsuite/MDAnalysisTests/data/waterbridge`. Then everything should work as for other classes, I guess.

talagayev commented:
> @talagayev I don't see a reason why these files take `StringIO` instead of just a filename, hence I'd suggest re-writing these tests to accept filenames, and put respective filenames (`ADA.gro`, `DWA.gro`, ...) in e.g. `testsuite/MDAnalysisTests/data` or in a new folder `testsuite/MDAnalysisTests/data/waterbridge`. Then everything should work as for other classes, I guess.

Agree, we can do them similar to the other cases with the files. I will create the test filles, put them into  `testsuite/MDAnalysisTests/data/waterbridge` and adjust the pytests to use them. I will then do for now for the tests the case where I check for now for for each case a test with 1 frame, basically as they are now and a test for multiple frames, so similar like the added `test_donor_accepter_pbc_multi`, just name them more correspondingly to the tests with 1 frame.

talagayev committed:
adjusted the test files for WB

talagayev committed:
black formatting

talagayev committed:
removed files that were moved

talagayev committed:
black formatting

marinegor commented:
@talagayev ok, sure! just `@` me here or request a review when you're done.

talagayev committed:
fixed case when empty results with parallelization

talagayev commented:
@marinegor okay should be ready to be reviewed now, had to do adjust `timesteps` for parallelization cases, due to an error that was appearing, mainly that one that was appearing

```
analysis_func = <function TestWaterBridgeAnalysis.test_count_by_time_weight.<locals>.analysis at 0x000002322B345A80>, kwargs = {}, result = []

    def count_by_time(self, analysis_func=None, **kwargs):
        """Counts the number of water bridges per timestep.

        The counting behaviour can be adjusted by supplying analysis_func.
        See :ref:`wb_count_by_time` for details.

        Returns
        -------
        counts : list
             Returns a time series ``N(t)`` where ``N`` is the total
             number of observed water bridges at time ``t``.

        """
        if analysis_func is None:
            analysis_func = self._count_by_time_analysis
        if self.results.network:
            result = []
>           for time, frame in zip(self.timesteps, self.results.network):
                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E           TypeError: 'NoneType' object is not iterable

..\..\..\package\MDAnalysis\analysis\hydrogenbonds\wbridge_analysis.py:1966: TypeError
```

So had to adjust `timesteps` to not be `NoneType` during parallelization.

As for the testfiles, I did create those that were created in the tests, put them in the `data` and import them similar as in the other tests.

marinegor committed:
remove unused imports

talagayev commented:
@marinegor Ah and I can update the `CHANGELOG `and add the Documentation if this looks good, wasn't sure if something needs to be modified so I kept it for last :)

marinegor commented:
@talagayev 

I'm slightly worried about these timesteps -- shouldn't they be actual timestamps, and not just consecutive integers?

I.e. if I modify lines 1954-ish it to:

```python
        if self.results.network:
            result = []
            print(f'{self.timesteps=}')
            assert False
            for time, frame in zip(self.timesteps, self.results.network):
```

and run tests, I get `[0.0]` before assertion error.

I guess the issue here is that attribute `timesteps` are implicitly expected to exist after the run is finished, which is not the case when we run it in parallel. Hence they should be moved to `self.results.timesteps` if we rely on them in `_conclude` or elsewhere.

It's weirdly the only class that actually does this:

```
❯ rg -t py self.timesteps
package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py
1027:        self.timesteps = None  # time for each frame
1314:        self.timesteps = []
1409:        self.timesteps.append(self._ts.time)
1969:        if self.timesteps is None:
1972:            timesteps = self.timesteps
2036:            if self.timesteps is None:
2039:                timesteps = self.timesteps
2140:        for t, hframe in zip(self.timesteps, timeseries):
```

@MDAnalysis/coredevs do you have any opinion on moving `timesteps` to `results.timesteps` instead? If I'm understanding correctly, the right way to do that is to deprecate `self.timesteps` in favor of `self.results.timesteps`, while for now copying `self.results.timesteps` to `timesteps`.

talagayev commented:
> @talagayev
> 
> I'm slightly worried about these timesteps -- shouldn't they be actual timestamps, and not just consecutive integers?
> 
> I.e. if I modify lines 1954-ish it to:
> 
> ```python
>         if self.results.network:
>             result = []
>             print(f'{self.timesteps=}')
>             assert False
>             for time, frame in zip(self.timesteps, self.results.network):
> ```
> 
> and run tests, I get `[0.0]` before assertion error.
> 
> I guess the issue here is that attribute `timesteps` are implicitly expected to exist after the run is finished, which is not the case when we run it in parallel. Hence they should be moved to `self.results.timesteps` if we rely on them in `_conclude` or elsewhere.
> 
> It's weirdly the only class that actually does this:
> 
> ```
> ❯ rg -t py self.timesteps
> package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py
> 1027:        self.timesteps = None  # time for each frame
> 1314:        self.timesteps = []
> 1409:        self.timesteps.append(self._ts.time)
> 1969:        if self.timesteps is None:
> 1972:            timesteps = self.timesteps
> 2036:            if self.timesteps is None:
> 2039:                timesteps = self.timesteps
> 2140:        for t, hframe in zip(self.timesteps, timeseries):
> ```
> 
> @MDAnalysis/coredevs do you have any opinion on moving `timesteps` to `results.timesteps` instead? If I'm understanding correctly, the right way to do that is to deprecate `self.timesteps` in favor of `self.results.timesteps`, while for now copying `self.results.timesteps` to `timesteps`.

True, yes was not sure about the solution, if it will work for other cases, that are not covered by the pytests 🙈 Yes should be timesteps, which is currently tricky to get during multiprocessing, which is as you mentioned that they have to exist after the run, which works well with serial, but difficult with multiprocessing.

hm, yes the moving to of `timesteps` to  `results.timesteps` could be an option I think, would need to look more into it as well.

marinegor commented:
I vaguely remember having this conversation already, actually.

I think #3720 and #3735 (reading changelog) implemented something like that, so we could in principle do this right away too🫣 But I need other coderev opinions.

talagayev commented:
> I vaguely remember having this conversation already, actually.
> 
> I think #3720 and #3735 (reading changelog) implemented something like that, so we could in principle do this right away too🫣 But I need other coderev opinions.

Yep, sounds good, let's see what the others say, in the meantime I can check the PR and how it was done there.

orbeckst commented:
`timesteps` is not part of the [AnalysisBase API](https://docs.mdanalysis.org/stable/documentation_pages/analysis/base.html) (as far as I can tell) so I don't see an issue putting it under `results`. 

If it was documented as a top level attribute of the specific class before then you have to deprecate it and keep a pointer around until we can remove it in 3.0.

(Apologies, quick comment before I am now starting my 🦃 days.)

IAlibay commented on testsuite/MDAnalysisTests/analysis/test_wbridge.py:
```diff
@@ -12,268 +11,33 @@
-    4ALA      H    4   0.500   0.000   0.000
-    4ALA      N    5   0.600   0.000   0.000
-  1.0   1.0   1.0"""
-        u = MDAnalysis.Universe(StringIO(grofile), format="gro")
```
I'm not sure I follow the whole StringIO thing - but being able to read a StringIO is a core component of MDAnalysis. Are we saying that this is no longer possible?

talagayev replied:
The problem in the current iteration is that when StringIO is combined with parallelization it led to infinite pytests, so reading StringIO + doing analysis with `dask`/`multiprocessing` is not compatible. Maybe an option would to add then an exception case and some check before the running of an analysis if the file was read from StringIO 🤔

marinegor replied:
@talagayev can you perhaps provide a minimal running example of what was happening with `StringIO`, so that we could try running it locally and see the infiniteness for ourselves?

talagayev replied:
yes, I can try maybe to do a small push, put one StringIO case back so that you can then git clone it and run/test locally. Can do it in around 30 minutes.

talagayev replied:
okay got the `test_loop` to the intial state where it reads the `StringIO` to create the universe. It currently fails the tests here due to the time limit, so you can git clone and run it locally

marinegor replied:
for anyone wondering, here's the standalone code:

```python
from io import StringIO

import MDAnalysis as mda
from MDAnalysis.analysis.hydrogenbonds.wbridge_analysis import (
    WaterBridgeAnalysis,
)

s = """Test gro file
5
    1ALA      O    1   0.000   0.001   0.000
    2SOL     OW    2   0.300   0.001   0.000
    2SOL    HW1    3   0.200   0.002   0.000
    2SOL    HW2    4   0.200   0.000   0.000
    4ALA      O    5   0.600   0.000   0.000
  1.0   1.0   1.0"""
grofile = StringIO(s)


wb = WaterBridgeAnalysis(
    mda.Universe(grofile, format="gro"),
    "protein and (resid 1)",
    "protein and (resid 1 or resid 4)",
)
wb.run(backend="multiprocessing")

```

and here's the stacktrace:

```
In [9]: wb.run(backend='multiprocessing')
/home/marinegor/.local/share/uv/python/cpython-3.13.7-linux-x86_64-gnu/lib/python3.13/multiprocessing/reduction.py:51: UserWarning: Reader has no dt information, set to 1.0 ps
  cls(buf, protocol).dump(obj)
Process ForkPoolWorker-1:
Traceback (most recent call last):
  File "/home/marinegor/.local/share/uv/python/cpython-3.13.7-linux-x86_64-gnu/lib/python3.13/multiprocessing/process.py", line 313, in _bootstrap
    self.run()
    ~~~~~~~~^^
  File "/home/marinegor/.local/share/uv/python/cpython-3.13.7-linux-x86_64-gnu/lib/python3.13/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
    ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/marinegor/.local/share/uv/python/cpython-3.13.7-linux-x86_64-gnu/lib/python3.13/multiprocessing/pool.py", line 114, in worker
    task = get()
  File "/home/marinegor/.local/share/uv/python/cpython-3.13.7-linux-x86_64-gnu/lib/python3.13/multiprocessing/queues.py", line 387, in get
    return _ForkingPickler.loads(res)
           ~~~~~~~~~~~~~~~~~~~~~^^^^^
  File "/home/marinegor/github/mdanalysis/package/MDAnalysis/lib/util.py", line 736, in __getattr__
    return getattr(self.stream, x)
                   ^^^^^^^^^^^
  File "/home/marinegor/github/mdanalysis/package/MDAnalysis/lib/util.py", line 736, in __getattr__
    return getattr(self.stream, x)
                   ^^^^^^^^^^^
  File "/home/marinegor/github/mdanalysis/package/MDAnalysis/lib/util.py", line 736, in __getattr__
    return getattr(self.stream, x)
                   ^^^^^^^^^^^
  [Previous line repeated 2970 more times]
RecursionError: maximum recursion depth exceeded
Exception ignored in: <function NamedStream.__del__ at 0x70e968095800>
Traceback (most recent call last):
  File "/home/marinegor/github/mdanalysis/package/MDAnalysis/lib/util.py", line 736, in __getattr__
    return getattr(self.stream, x)
  File "/home/marinegor/github/mdanalysis/package/MDAnalysis/lib/util.py", line 736, in __getattr__
    return getattr(self.stream, x)
  File "/home/marinegor/github/mdanalysis/package/MDAnalysis/lib/util.py", line 736, in __getattr__
    return getattr(self.stream, x)
  [Previous line repeated 997 more times]
RecursionError: maximum recursion depth exceeded
^CProcess ForkPoolWorker-2:
```


@talagayev I'd suggest to make a separate issue out of it and not include it in this PR.

IAlibay replied:
> not include it in this PR

If you decide to move everything out of the StringIO representation, I would ask that you open an issue to fix it once the underlying issue is fixed.

orbeckst replied:
I'd prefer if we didn't need to add all these files and could continue using the in-line test snippets.

More broadly, before working around issues we should understand what's going wrong.

talagayev committed:
got back StringIO case for showcase

talagayev committed:
Merge branch 'develop' into wb_parallel

talagayev committed:
Update test_wbridge.py

get one StringIO case back for showcase

talagayev committed:
Update conftest.py

add missing

talagayev committed:
Update test_wbridge.py

import addition

marinegor commented:
also @talagayev you can have a look at `InterRDS_s` and `DeprecationWarning` there to have a look at how things get moved to `results`. I'd do something similar here as well.

marinegor commented on package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:
```diff
@@ -1951,25 +1961,32 @@ def count_by_time(self, analysis_func=None, **kwargs):
+        if self.timesteps is None:
+            timesteps = range(len(self.results.network))
+        else:
+            timesteps = self.timesteps
```
should be actual timesteps, and not integer indices. Solution would be to save them to `results._timesteps` and later use them where needed, while adding deprecation warning to `self.timesteps` alike `InterRDS_s`'s approach.

talagayev replied:
adjusted the code to use more `results._timesteps`. @marinegor could you take a look if it is correct now? :)

IAlibay commented on testsuite/MDAnalysisTests/datafiles.py:
```diff
@@ -392,8 +392,24 @@
     "SURFACE_PDB",  # 111 FCC lattice topology for NSGrid bug #2345
     "SURFACE_TRR",  # full precision coordinates for NSGrid bug #2345
     "DSSP",  # DSSP test suite
+    "WB_AD",
```
Please add desriptors for all these files, the filenames themselves won't make a ton of sense a couple months down the line.

talagayev replied:
Added descriptions for the files using the descriptions that were used in the `StringIO` files

IAlibay commented on testsuite/MDAnalysisTests/analysis/conftest.py:
```diff
@@ -217,3 +220,11 @@ def client_InterRDF_s(request):
     return request.param
+
+  
+# MDAnalysis.analysis.hydrogenbonds.wbridge_analysis
```
remove?

talagayev replied:
Can remove, but would it not better to keep it for consistency with the other functions, that are also commented in that way like here:

https://github.com/talagayev/mdanalysis/blob/5871f5c876bc873c19f9fc8c9694e40c1f2021a7/testsuite/MDAnalysisTests/analysis/conftest.py#L217-L222

orbeckst replied:
Would keep for consistency.

orbeckst commented:
@talagayev why is StringIO not working here?

marinegor commented:
I pasted traceback above -- seems that StringIO doesn't work well with multiprocessing.

talagayev commented:
> I pasted traceback above -- seems that StringIO doesn't work well with multiprocessing.

Could it be connected somehow to the case, that at some point in the parallelization if it uses `ReaderBase` to read the file it would try to read the file and would not be able to access it, since it gets only a string that is not pickable and iteratable due to this part of the code in the `coordinates/base.py`? Since here he only exception being `NamedStream` files.

```
    if isinstance(filename, NamedStream):
        self.filename = filename
    else:
        self.filename = str(filename)
```

maybe we can adjust it somehow for `StringIO` cases. How is it currently regulated with `NamedStream` cases? I assume those would possibly also fall into the same line of them being difficult to parallelize?

marinegor commented:
@orbeckst @IAlibay do you have any feedback on moving `StringIO`-based tests for this PR?

I honestly see this as a separate issue, which we totally must raise (@talagayev please tell me if you wanna do that, otherwise I'll do it). But, I don't see why it would block this particular PR enabling parallelization.

@talagayev could you please implement all non-StringIO related changes requested by me and Irfan and re-request review when it's ready?

orbeckst commented:
Do you suggest to make the tests here run with file-based access and then revert to StringIO when it's fixed?

marinegor commented:
I generally do not understand why particularly here tests are StringIO-based -- it doesn't seem to be that common afaik.

I'd suggest making them file-based (as the rest of the tests) and raising an issue about using StringIO topology/universe with non-serial backends.

orbeckst commented:
StringIO cuts down on the number of files to ship and it's neat when you only have mini-files that you can keep in-line.

orbeckst reviewed (commented):
Although I prefer the StringIO-based testing here, I am ok with moving forward with the file-based workaround, provided you raise an issue and add comments to tests.filenames referencing this PR, the new issue, and a note that we may want to turn files back into in-line when the issue about StringIO/multiprocessing/pickling has been resolved.

marinegor commented:
I agree, it reduces numbers of files indeed, and is more readable. But in this particular case I feel like passing tests for the class that actually should run on real-world data (=files) could have a higher priority. I'm not sure if StringIO is ever used in real-world scenarios, and it feels weird to me to block this PR because of the (inconsistent with other AnalysisBase subclasses) way tests are written here.

And indeed as you mentioned, we can indeed revert to StringIO when the issue with it in parallel context is resolved.

IAlibay commented:
@marinegor StringIO is very frequently used in the real world, please do not discount it as a second class citizen, otherwise a chunk of the ecosystem is going to fall apart.

marinegor commented:
Sorry if I came across a bit too strong, I don't have problems with StringIO in general. But in this particular case, seems that

 - StringIO doesn't work for any non-serial run for any AnalysisBase subclass
 - this particular PR is correct, and fails because of the issue above

Hence I suggest not fixing StringIO issue here but rather raising it separately, while accepting this PR (and switching it to file-based testing).

talagayev commented:
> @orbeckst @IAlibay do you have any feedback on moving `StringIO`-based tests for this PR?
> 
> I honestly see this as a separate issue, which we totally must raise (@talagayev please tell me if you wanna do that, otherwise I'll do it). But, I don't see why it would block this particular PR enabling parallelization.
> 
> @talagayev could you please implement all non-StringIO related changes requested by me and Irfan and re-request review when it's ready?

Yup can do, but will quickly wait for an decision how we proceed with `StringIO` and can then adjust the code and create the Issue.

talagayev committed:
Merge branch 'MDAnalysis:develop' into wb_parallel

talagayev committed:
suggested PR adjustment

talagayev committed:
modified test

talagayev committed:
Merge branch 'wb_parallel' of https://github.com/talagayev/mdanalysis into wb_parallel

talagayev committed:
modified test

talagayev committed:
black formatting

talagayev committed:
black formatting

talagayev committed:
Update datafiles.py

addition of comments to the files

talagayev committed:
comments + black formatting

talagayev commented:
I created the Issue #5221 that references the problem of the `StringIO` with the parallelization backends, added comments to the files in`datafiles.py` that I retrieved from the `StringIO` tests to specify what every file contains and added notes about the usage of files instead fo `StringIO` objects for now with in the future reverting back to `StringIO` when the Issues are adressed.

For the notes I am not sure if the ones I added are enough or if I should add additional ones.

orbeckst commented on package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:
```diff
@@ -1395,8 +1406,37 @@ def _donor2acceptor(self, donors, h_donors, acceptor):
+        normalizes such cases and falls back to frame indices.
+        """
+        n = len(self.results.network)
+        ts = self.results.timesteps
```
I would call this variable something else — we use `ts` everywhere in the code to mean "an instance of `Timestep`, add it's quite confusing when it doesn't mean that.

Just call it `timesteps` (which also includes the plural to indicate that this is a container).

orbeckst commented on package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:
```diff
@@ -1395,8 +1406,37 @@ def _donor2acceptor(self, donors, h_donors, acceptor):
+            # if only one frame, accept scalar; otherwise fall back
+            return [item] if n == 1 else range(n)
+
+        # empty or mismatched length -> fall back
```
Is this also related to scalars?? Document in doc string?

orbeckst commented on package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:
```diff
@@ -1395,8 +1406,37 @@ def _donor2acceptor(self, donors, h_donors, acceptor):
             )
         return result
 
+    def _iter_timesteps(self):
```
This looks like a hack/band-aid. I believe you that it's necessary. Just make sure that you give as much information in the doc string/comments so that it's clear why it's here.

Is this code tested??

orbeckst commented on package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:
```diff
@@ -1014,7 +1024,8 @@ def __init__(
         self.results.timeseries = None
-        self.timesteps = None  # time for each frame
+        self.results.timesteps = None
+        self._timesteps = []
```
Do we need `results.timesteps` and `_timesteps`? How does this differ from the standard [`times`](https://docs.mdanalysis.org/stable/documentation_pages/analysis/base.html#MDAnalysis.analysis.base.AnalysisBase.times) that come with AnalysisBase

orbeckst commented on package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:
```diff
@@ -1395,8 +1406,37 @@ def _donor2acceptor(self, donors, h_donors, acceptor):
+
     def _single_frame(self):
-        self.timesteps.append(self._ts.time)
+        self._timesteps.append(self._ts.time)
```
This replicates standard `AnalysisBase.times` or am I missing something?

orbeckst commented on package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:
```diff
@@ -2171,3 +2251,12 @@ def timeseries(self):
+    def timesteps(self):
+        wmsg = (
+            "The `timesteps` attribute is deprecated and will be removed in "
+            "MDAnalysis 3.0.0. Please use `results.timesteps` instead."
```
Recommend to use `times` (assuming it's equivalent)?

orbeckst commented on package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:
```diff
@@ -2150,8 +2218,20 @@ def generate_table(self, output_format=None):
     def _conclude(self):
+        # saving timesteps in results for parallelization
+        self.results.timesteps = np.asarray(self._timesteps)
+
```
remove empty line?

orbeckst reviewed (changes requested):
Thanks for the ton of work you've been doing @talagayev . I did not deep-dive into the code so I may be getting things wrong. 

1. What is the issue with mismatches between recorded times and results? Is it the case that there are some frames that do not have a water bridge?

   Does this only create a problem in parallel, and if so, why?
 
   At the moment you're addressing the issue by adjusting the list of times. Could you alternatively address it by recording each analyzed frames (basically, what AnalysisBase already provides in [`AnalysisBase.times`](https://docs.mdanalysis.org/stable/documentation_pages/analysis/base.html#MDAnalysis.analysis.base.AnalysisBase.times) and instead record `None` for a frame without a water bridge? Then one had a 1-t-1 mapping frames <-> results throughout and can decide how to clean up in `_conclude` (once we see all data).

   This patch is not covered by tests (see [codecov](https://app.codecov.io/gh/MDAnalysis/mdanalysis/pull/5151?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=checks&utm_campaign=pr+comments&utm_term=MDAnalysis)) so we definitely need something to demonstrate that it works as intended.
2. The code is using `timesteps` to record the list of times. My understanding is that [`AnalysisBase.times`](https://docs.mdanalysis.org/stable/documentation_pages/analysis/base.html#MDAnalysis.analysis.base.AnalysisBase.times) is doing exactly that already so we should just use [`AnalysisBase.times`](https://docs.mdanalysis.org/stable/documentation_pages/analysis/base.html#MDAnalysis.analysis.base.AnalysisBase.times) and schedule `timesteps` for removal (deprecate as you're already doing, but don't even add to `results`).

Thank you for raising #5221. It could be convenient if you rewrote the history of this PR in such a way that there's one commit in which StringIO was removed and files added because then, when #5221 is fixed, one could just revert that commit. If that's too much work then don't worry.

talagayev commented:
> Thanks for the ton of work you've been doing @talagayev . I did not deep-dive into the code so I may be getting things wrong.
> 
> 1. What is the issue with mismatches between recorded times and results? Is it the case that there are some frames that do not have a water bridge?
>    Does this only create a problem in parallel, and if so, why?
>    At the moment you're addressing the issue by adjusting the list of times. Could you alternatively address it by recording each analyzed frames (basically, what AnalysisBase already provides in [`AnalysisBase.times`](https://docs.mdanalysis.org/stable/documentation_pages/analysis/base.html#MDAnalysis.analysis.base.AnalysisBase.times) and instead record `None` for a frame without a water bridge? Then one had a 1-t-1 mapping frames <-> results throughout and can decide how to clean up in `_conclude` (once we see all data).
>    This patch is not covered by tests (see [codecov](https://app.codecov.io/gh/MDAnalysis/mdanalysis/pull/5151?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=checks&utm_campaign=pr+comments&utm_term=MDAnalysis)) so we definitely need something to demonstrate that it works as intended.
> 2. The code is using `timesteps` to record the list of times. My understanding is that [`AnalysisBase.times`](https://docs.mdanalysis.org/stable/documentation_pages/analysis/base.html#MDAnalysis.analysis.base.AnalysisBase.times) is doing exactly that already so we should just use [`AnalysisBase.times`](https://docs.mdanalysis.org/stable/documentation_pages/analysis/base.html#MDAnalysis.analysis.base.AnalysisBase.times) and schedule `timesteps` for removal (deprecate as you're already doing, but don't even add to `results`).
> 
> Thank you for raising #5221. It could be convenient if you rewrote the history of this PR in such a way that there's one commit in which StringIO was removed and files added because then, when #5221 is fixed, one could just revert that commit. If that's too much work then don't worry.

Thank you @orbeckst 😃 yes here the case why it happens is a quite specific case, where there are more workers assigned then there are frames to be analyzed, which resulted in the problem that worker that didn't have a frame to work with gave nothing and this led to the error. Yes true, this is a problem that does not occur with something like `RMSD` and the other analysis functions, so the best approach would be the one you suggested will adjust it accordingly, somehow didn't think about that 🙈 will adjust everything accordingly and ping when it is ready, should take a couple of days during this week.

talagayev committed:
removal of teststeps and use of times

talagayev committed:
black format

talagayev committed:
adjusted tests

talagayev committed:
adjusted code

talagayev committed:
added versionchanged

talagayev committed:
added CHANGELOG entry

talagayev committed:
Merge branch 'develop' into wb_parallel

talagayev commented:
> > Thanks for the ton of work you've been doing @talagayev . I did not deep-dive into the code so I may be getting things wrong.
> > 
> > 1. What is the issue with mismatches between recorded times and results? Is it the case that there are some frames that do not have a water bridge?
> >    Does this only create a problem in parallel, and if so, why?
> >    At the moment you're addressing the issue by adjusting the list of times. Could you alternatively address it by recording each analyzed frames (basically, what AnalysisBase already provides in [`AnalysisBase.times`](https://docs.mdanalysis.org/stable/documentation_pages/analysis/base.html#MDAnalysis.analysis.base.AnalysisBase.times) and instead record `None` for a frame without a water bridge? Then one had a 1-t-1 mapping frames <-> results throughout and can decide how to clean up in `_conclude` (once we see all data).
> >    This patch is not covered by tests (see [codecov](https://app.codecov.io/gh/MDAnalysis/mdanalysis/pull/5151?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=checks&utm_campaign=pr+comments&utm_term=MDAnalysis)) so we definitely need something to demonstrate that it works as intended.
> > 2. The code is using `timesteps` to record the list of times. My understanding is that [`AnalysisBase.times`](https://docs.mdanalysis.org/stable/documentation_pages/analysis/base.html#MDAnalysis.analysis.base.AnalysisBase.times) is doing exactly that already so we should just use [`AnalysisBase.times`](https://docs.mdanalysis.org/stable/documentation_pages/analysis/base.html#MDAnalysis.analysis.base.AnalysisBase.times) and schedule `timesteps` for removal (deprecate as you're already doing, but don't even add to `results`).
> > 
> > Thank you for raising #5221. It could be convenient if you rewrote the history of this PR in such a way that there's one commit in which StringIO was removed and files added because then, when #5221 is fixed, one could just revert that commit. If that's too much work then don't worry.
> 
> Thank you @orbeckst 😃 yes here the case why it happens is a quite specific case, where there are more workers assigned then there are frames to be analyzed, which resulted in the problem that worker that didn't have a frame to work with gave nothing and this led to the error. Yes true, this is a problem that does not occur with something like `RMSD` and the other analysis functions, so the best approach would be the one you suggested will adjust it accordingly, somehow didn't think about that 🙈 will adjust everything accordingly and ping when it is ready, should take a couple of days during this week.

@orbeckst looks like the main problem with the parallelization was basically overall the use of `timesteps`, since completly removing `timesteps` and opting instead for `AnalysisBase.times` works well with the parallelization. I adjusted it, which is basically removing `timesteps` and adjusting the code to use `AnalysisBase.times` and then we don't require some go-arounds to make the parallelization work :)

If we want to take that approach I wanted to ask what the procedure is now in that case with `timesteps` being removed, since they are not used? A warning that now `timesteps` are replaced by `AnalysisBase.times` etc.?
